job-runner/server/server.go

```diff
@@ -55,6 +108,60 @@ func (s *Server) Boot() error {
 	return nil
 }
 
+// Stop stops the server. It signals running traversers to shut down and then
+// stops the API (using either the default api.Stop or the StopAPI hook if
+// provided). Once Stop has been called, the server cannot be reused - future calls
+// to Run will return an error.
+//
+// If autoShutdown was set when calling Run, Stop will automatically be called by
+// the server on receiving a TERM or INT signal from the OS. Otherwise, you must
+// call Stop when you want to shut down the Job Runner.
+func (s *Server) Stop() error {
+	// Only stop once. We lock the whole Stop call, so that, if Stop is called
+	// multiple times in quick succession, no calls will return before the server
+	// has actually been shut down.
+	s.stopMux.Lock()
+	defer s.stopMux.Unlock()
+	if s.stopped {
+		return nil
+	}
+	s.stopped = true
+
+	log.Infof("Stopping Job Runner server")
+
+	// Running traversers watch shutdownChan - closing this tells them to shut down.
+	// The API will also begin refusing to start running new job chains.
+	close(s.shutdownChan)
+
+	// Wait for all traversers to shut down. Timeout if they aren't done
+	// within 20 seconds, and continue to shutting down the API.
+	timeout := time.After(20 * time.Second)
+WAIT_FOR_TRAVERSERS:
+	for !s.traverserRepo.IsEmpty() {
+		select {
+		case <-time.After(10 * time.Millisecond):
+			// Check again if traversers are all done.
+		case <-timeout:
+			break WAIT_FOR_TRAVERSERS
+		}
+	}
+
+	// Stop the API, using the StopAPI hook if provided and api.Stop otherwise.
```

Comment:
Do we need to stop API first? I.e. so it won't accept any new work.

Reply:
When we close the shutdownChan (line 134), the API stops accepting new job chains to run (also SJCs to resume). There's a comment there to that effect.

Reply:
👍